### PR TITLE
CORE-6732, CORE-6688: new test cases

### PR DESCRIPTION
**Pull request by dickon**
CORE-6625: add new test cases, and some test case refactoring
Some helper functions which can later be used from other classes are now in a separate helpers file.

### CODE REVIEW

**Comment by dickon** [comment]
I've added `equals` and `hashCode` where necessary; maybe we should consistently implement those methods for all our entity classes which are not `data class`.
 
There's still some scope for tidying up the test cases a little, e.g. eliminate makeCpks.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkMetadataEntity.kt [hunk 1]
```diff
@@ -28,16 +29,37 @@ data class CpkMetadataEntity(
...
+    override fun equals(other: Any?) = other is CpkMetadataEntity &&
+            other.id == id &&
+            other.cpkFileChecksum == cpkFileChecksum &&
+            other.formatVersion == formatVersion &&
+            other.serializedMetadata == serializedMetadata &&
+            other.isDeleted == isDeleted
+    // deliberately ignore entityVersion since that's not really part of the value
+
+    override fun hashCode() =
+        id.hashCode() + cpkFileChecksum.hashCode() + formatVersion.hashCode() + serializedMetadata.hashCode() + isDeleted.hashCode()
```
Comment (sprintengine) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkMetadataEntity.kt]: Just use the ID and cpkFIleChecksum as these are unique identifiers for instances of this table (maybe you don't even need cpkFileChecksum)

```
override fun equals(other: Any?) = other is CpkMetadataEntity &&
            other.id == id &&
            other.cpkFileChecksum == cpkFileChecksum

override fun hashCode() =
        id.hashCode() + cpkFileChecksum.hashCode()
```

Source - https://thorben-janssen.com/ultimate-guide-to-implementing-equals-and-hashcode-with-hibernate/#:~:text=Java's%20default%20implementation%20of%20the,entity%20twice%20within%20a%20Session.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt [hunk 2]
```diff
@@ -19,11 +19,12 @@ import javax.persistence.Version
 /**
  * Cpi/cpk mapping table.
  */
+@Suppress("LongParameterList")
```
Comment (sprintengine) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt]: Is this actually needed here?

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt [hunk 3]
```diff
@@ -42,32 +43,60 @@ data class CpiCpkEntity(
...
-
+) {
     @PreUpdate
     fun onUpdate() {
         insertTimestamp = Instant.now()
     }
+
+    override fun equals(other: Any?): Boolean = other is CpiCpkEntity &&
+            other.id == id &&
+            other.cpkFileName == cpkFileName &&
```
Comment (sprintengine) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt]: Same here, just the id comparison necessary and generate hashcode using id.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt [hunk 4]
```diff
@@ -42,32 +43,60 @@ data class CpiCpkEntity(
...
     @Column(name = "cpk_version")
-    private val cpkVersion: String,
+    var cpkVersion: String,
     @Column(name = "cpk_signer_summary_hash")
-    private val cpkSignerSummaryHash: String,
-): Serializable
\ No newline at end of file
+    var cpkSignerSummaryHash: String,
+) : Serializable {
+    override fun equals(other: Any?): Boolean = other is CpiCpkKey &&
```
Comment (sprintengine) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt]: Same here, just the id comparison necessary and generate hashcode using id.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiMetadataEntity.kt [hunk 5]
```diff
@@ -118,22 +81,27 @@ data class CpiMetadataEntity(
...
+    ) = CpiMetadataEntity(
+        id,
+        fileName,
+        fileChecksum,
+        groupPolicy,
+        groupId,
+        fileUploadRequestId,
+        cpks
+    )
 }
```
Comment (sprintengine) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiMetadataEntity.kt]: Think this missing equals and hashcode

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkDbChangeLogEntity.kt [hunk 6]
```diff
@@ -21,31 +21,36 @@ class CpkDbChangeLogEntity(
...
 
     // this TS is managed on the DB itself
     @Column(name = "insert_ts", insertable = false, updatable = false)
-    val insertTimestamp: Instant? = null
+    var insertTimestamp: Instant? = null
+
+    @PreUpdate
+    fun onUpdate() {
+        insertTimestamp = Instant.now()
+    }
```
Comment (sprintengine) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkDbChangeLogEntity.kt]: This also missing equals and hashcode

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkFileEntity.kt [hunk 7]
```diff
@@ -49,15 +52,15 @@ data class CpkFileEntity(
...
-    var entityVersion: Int = 0
-
+    var entityVersion: Int = 0,
     // this TS is managed on the DB itself
     @Column(name = "insert_ts", insertable = false, updatable = false)
     var insertTimestamp: Instant? = null
+) {
+
 
     override fun equals(other: Any?): Boolean {
```
Comment (sprintengine) [inline libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkFileEntity.kt]: I suspect this equals and hashcode might need changed too, they include the data which is not part of the primary key. Maybe this could lead to some strange behaviour if we put CpkFileEntities into a set or hashmap, or if we have lazy loaded the instance (with no data), or if we have updated an entity's data and have some detached entities.

Might be worth removing data from the equals and hashcode

**Review by conal-smith — COMMENTED** [review]
Just a few comments on equals and hashcode implementations.

Hibernate docs say you should override these if you intend to put managed instances into a set or use reattachment of detached entities, the latter I don't think we do anyway, but we do have CpiCpkEntity in a set so this one we should add the overrides with the identifier.

- https://docs.jboss.org/hibernate/core/4.0/manual/en-US/html/persistent-classes.html#persistent-classes-equalshashcode

TL;DR, If you compare using the primary key, then if two entities are equal, they are the same row, and I think that's what you want with JPA entities and equality. If both are added to a set, you'll just have 1 element in the set.

**Comment by vkolomeyko** [comment]
Given that we have https://github.com/corda/corda-runtime-os/pull/2172 in progress being worked on by @benmcmahon100 - we need to decide in which order we are going to merge those 2 PRs.

**Comment by dickon** [comment]
This PR rotted quite badly. After agreement with @vkolomeyko , @benmcmahon100, @conalsmith-r3 let's try and get this work through in small pieces. Therefore, this first PR now just has one file of test case changes.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkDbChangeLogEntity.kt (hunk 6 above)
Comment (dickon) [reply libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkDbChangeLogEntity.kt]: That change is now slated for another PR.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiMetadataEntity.kt (hunk 5 above)
Comment (dickon) [reply libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiMetadataEntity.kt]: That change is now slated for another PR.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt (hunk 4 above)
Comment (dickon) [reply libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt]: That change is now slated for another PR.

I've been doing some more reading on this, and the situation is more nuanced, but you are right in that's all we need in this case.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt (hunk 3 above)
Comment (dickon) [reply libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt]: That change is now slated for another PR.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt (hunk 2 above)
Comment (dickon) [reply libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt]: That change is now slated for another PR. It was needed at one point.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkMetadataEntity.kt (hunk 1 above)
Comment (dickon) [reply libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkMetadataEntity.kt]: There's a nice detailed discussion of the issue here 

https://web.archive.org/web/20170710132916/http://www.onjava.com/pub/a/onjava/2006/09/13/dont-let-hibernate-steal-your-identity.html?page=3

which I found via the equally useful

https://stackoverflow.com/questions/5031614/the-jpa-hashcode-equals-dilemma

Anyway, this is for another PR now.

File: components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt [hunk 8]
```diff
@@ -769,24 +467,29 @@ internal class DatabaseCpiPersistenceTest {
...
+
+    @Test
+    fun `can store multiple versions of the same CPI name in the same group`() {
+        val name = UUID.randomUUID().toString()
+        val cpi_v1 = mockCpi(mockCpk(), name = name, version = "v1")
+        val cpi_v2 = mockCpi(mockCpk(), name = name, version = "v2")
+        cpiPersistence.storeWithTestDefaults(cpi_v1)
+        cpiPersistence.storeWithTestDefaults(cpi_v2)
+        // this shouldn't pass but does
+    }
```
Comment (benmcmahon100) [inline components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt]: This doesn't look good.

IMO I think it'd make more sense to add the assert that requires it fails, and mark it as 
`@Disabled("pending https://link/to.ticket)`

**Review by benmcmahon100 — DISMISSED** [review]
Only one small concern on my part. My main worry is we'll forget about this.

**Review by conal-smith — DISMISSED** [review]
LGTM, @benmcmahon100 the test you refer to above is now disabled as per 

```
@Disabled("https://r3-cev.atlassian.net/browse/CORE-7016")
```

File: components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt [hunk 9]
```diff
@@ -769,24 +468,29 @@ internal class DatabaseCpiPersistenceTest {
...
-        changelogsWithout.forEach {
-            println("${it.id} ${it.isDeleted}")
-        }
         assertThat(changelogsWithout.size).isEqualTo(0)
     }
 
+
+    @Disabled("https://r3-cev.atlassian.net/browse/CORE-7016")
+    @Test
+    fun `cannot store multiple versions of the same CPI name in the same group`() {
```
Comment (sprintengine) [inline components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt:481]: I'm a little confused about this test, are you sure it's valid? If a group cannot have two CPIs with the same name but different versions, how will a virtual node in the group ever be able to upgrade their CPI to the new version? The CPI has to exist already. Hence why there is no uniqueness constraint on (name, group_id) in `cpi` table.

File: libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkFileEntity.kt (hunk 7 above)
Comment (dickon) [reply libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpkFileEntity.kt]: To be address in a new PR.

File: components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt (hunk 9 above)
Comment (dickon) [reply components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt:481]: The original bug and discussion was https://r3-cev.atlassian.net/browse/CORE-6068 ; I have updated the reference in the @Disabled annotation.

Comment (dickon) [reply components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt:481]: And, to be clear, by the time update is ready and to be supported we should delete or rewrite this test case.

Comment (dickon) [reply components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt:481]: The point is that @driessamyn has said upgrade (without force) should not work in DP2, but the fact is the test case, before I added  the assertion that it should fail, passed fine due to missing checks in corda-runtime-os.

Comment (sprintengine) [reply components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt:481]: I do see there is validation in CpiValidatorImpl
```
cpiPersistence.verifyGroupIdIsUniqueForCpi(cpi)
```
This just loads the CPI by its full identifier (name, version, ssh) and gets the group, so it will always allow two CPIs with different versions in same group.

If it were to actually enforce uniqueness on (CPI name, groupId) it would need to find all CPIs with the CPI name via `em.find("FROM CpiMetadataEntity where name = :name"` and make sure the group is unique.

I think current behaviour is correct in terms of support for virtual node upgrade, which I'm actively designing and POC'ing and will be implementing in the coming weeks. 

Good to keep this in mind for my design. But for this PR I'm happy to have this merged and the referenced JIRA can decide what to do about that test.

Comment (dickon) [reply components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt:481]: Okay

File: components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt (hunk 8 above)
Comment (dickon) [reply components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt]: Done

**Review by conal-smith — APPROVED** [review]
LGTM

**Review by benmcmahon100 — APPROVED** [review]
